Add doc for subscribed events and events workflows

MVrachev opened the pull request:
It will be really useful if we add documentation for the
different events we are subscribed to answering the
questions "why we are subscribed to them" and
"when" they are triggered.

Closes: https://github.com/vmware/precaution/issues/211

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+
+## Subscribed events
+
+Precaution needs to subscribe to a number of events when it's
```
"when it's to satisfy...." - sounds awkward.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+Precaution needs to subscribe to a number of events when it's
+to satisfy different use cases of our users.
+
+The events can be separated into two groups `pull requests
```
"can be" or "are"?

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+
+### Pull requests events
+
+Official GitHub documentation: [PullRequest event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)
```
```suggestion
Official GitHub documentation: [Pull Request event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)
```

joshuagl replied:
The event name in the GitHub docs is `PullRequestEvent`, I favour using that as the link title

MVrachev replied:
I agree with Joshua.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+
+Official GitHub documentation: [PullRequest event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)
+
+* pull_request.opened - the first event which occurs when a new pull request is created. It's a self-explanatory
```
"It's a self-explanatory" - bad grammar.  Don't think this statement adds any value since you already explained it.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+* pull_request.reopened - notifies Precaution to start a new scan on a reopened pull request
+
+* pull_request.synchronize - occurs when you add a commit to already opened pull request.
+We discussed it and added it in pull request number [#113](https://github.com/vmware/precaution/pull/113)
```
Don't ever use "we" or "I" in documentation.  A reader doesn't know who "we" is.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+* pull_request.synchronize - occurs when you add a commit to already opened pull request.
+We discussed it and added it in pull request number [#113](https://github.com/vmware/precaution/pull/113)
+
+### Check run/suite events
```
Why "/"  and not just "and"?

MVrachev replied:
Okay, maybe you are right. I will change it.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+- [CheckRun event](https://developer.github.com/v3/activity/events/types/#checkrunevent)
+- [CheckSuite event](https://developer.github.com/v3/activity/events/types/#checksuiteevent)
+
+You had already opened a pull request and you are at the pull request page.
```
Avoid using "You"

MVrachev replied:
Maybe you are right. 
It will be better if I write from a third person view?

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+
+* check_suite.rerequested - happens when you click "Re-run all checks" 
+or "Re-run failed checks" buttons. 
+This notifies Precaution and the other GitHub apps 
```
What other GitHub apps?  The context here is unclear.

MVrachev replied:
Okay, I won’t mention the other GitHub apps at all.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+
+## Events workflows
+
+This is a small paragraph about the events workflow we are receiving with and posting to GitHub.
```
Delete this line.  No point to describing what you will describe.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+If you want to read more details how Precaution works you can look at our [Architecture]().
+
+There are numerous use cases but the triggered events and their order are alike.
+In the text below "F" equals "forked repository" and "U" means upstream repository.
```
Why use abbreviations?

MVrachev replied:
Because without them the titles of the different sections are too long. :D 
I will remove the abbreviations and you will tell me what you think.

joshuagl replied:
You've stopped using the abbreviations below, so no need to define them.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+5. `GET /repos/:owner/:repo/contents/:path` - multiple API calls to get the modified files
+6. `PATCH /repos/:owner/:repo/check-runs/:check_run_id` - API call which sends the results
+
+If you have already opened a pull request and you navigate to the "Checks" tab then
```
Run-on sentence

joshuagl replied:
This duplicates information from the Check*Event section, no need to repeat it here.

MVrachev replied:
Yea, I will remove it. Probably we don’t need it.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+
+This is a small paragraph about the events workflow we are receiving with and posting to GitHub.
+It doesn't aim to explain in details how Precaution works.
+If you want to read more details how Precaution works you can look at our [Architecture]().
```
Who is "our"???

joshuagl replied:
the link to the architecture diagram is missing

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+### Check run/suite events
+
+Official GitHub documentation:
+- [CheckRun event](https://developer.github.com/v3/activity/events/types/#checkrunevent)
```
```suggestion
+ [CheckRunEvent](https://developer.github.com/v3/activity/events/types/#checkrunevent)
```

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,89 @@
+
+Official GitHub documentation:
+- [CheckRun event](https://developer.github.com/v3/activity/events/types/#checkrunevent)
+- [CheckSuite event](https://developer.github.com/v3/activity/events/types/#checksuiteevent)
```
```suggestion
- [CheckSuiteEvent](https://developer.github.com/v3/activity/events/types/#checksuiteevent)
```

joshuagl commented:
Referring to my previous review, copied below:

> * let's standardise how we document events. For some we document what triggers the event, for some what happens when the event is triggered and for others we do both. We should list each event we respond to and for each of those summarise what triggers the event and why Precaution needs to listen to that event. For example:
>   > `pull_request` with `opened` status: triggered when a new pull request is opened on the repository. The initiating event for a Precaution workflow, starts a new scan on the pull request.

I still think we can do better here. Let's be consistent in terms of style and formatting.

MVrachev commented:
I updated my pr with the suggestions given by @joshuagl https://github.com/vmware/precaution/pull/241#pullrequestreview-270232476

I changed every event with the template you suggested and I made them bold.
What do you think @joshuagl?

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,103 @@
+
+*occurs - when a pull request is reopened from a closed state*
+
+*triggers: start of a new Precaution scan*
```
I feel like we need to delineate this triggers from `pull_request.opened`. "Re-start Precaution scan?"

joshuagl replied:
Or maybe we just need to be consistent in all of the triggers. All triggered events are effectively the same, right? scanning all changes in the PR

MVrachev replied:
Yes, all triggers are the same. The pull request events are just different entry point into the Precaution workflow - which is basically a scan on all files in the pull request.

MVrachev commented:
I updated my pr.
Because all `triggers` are actually different entry point to a new Precaution scan I changed them to be the same and I removed one redundant line.

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,102 @@
+## Subscribed events
+
+Precaution needs to subscribe to a number of events
+to satisfy different use cases of our users.
```
"of our users" is redundant here.

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,102 @@
+*occurs - when a new pull request is created*
+
+*triggers - start of the Precaution scan*
+
```
I'm still not a fan of the formatting choices here, what do you think of this?
#### `pull_request.opened`
- __occurs__: when a new pull request is created
- __triggers__: start of the Precaution scan

MVrachev replied:
I didn't know that markdown tricks.
A good idea.

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,102 @@
+
+*triggers - start of a new Precaution scan*
+
+*for more information look at pull request number [#113](https://github.com/vmware/precaution/pull/113)*
```
```suggestion
For more information look at pull request number [#113](https://github.com/vmware/precaution/pull/113)
```

Drop the italics and use sentence case

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,102 @@
+- [CheckRunEvent](https://developer.github.com/v3/activity/events/types/#checkrunevent)
+- [CheckSuiteEvent](https://developer.github.com/v3/activity/events/types/#checksuiteevent)
+
+A user had already opened a pull request and he is at the pull request page.
```
Drop redundant use of gendered pronoun
```suggestion
A user had already opened a pull request and is at the pull request page.
```

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,102 @@
+## Events workflows
+
+This paragraph doesn't aim to explain in details how Precaution works.
+If you want to read more details how Precaution works you can look at projects [Architecture](architecture.md).
```
This reads a bit awkwardly as there's repeated phrasing, I suggest we change the second sentence to something like

```suggestion
If you want to understand more details of Precaution's design see the [Architecture](architecture.md) documentation.
```

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,102 @@
+The events are separated into two groups `pull requests
+events` and `check suites/runs events`.
+
+### Pull requests events
```
I thought we'd agreed to use the event name from the GitHub docs, PullRequestEvent, as the section title?

joshuagl commented on docs/events_handling.md:
```diff
@@ -0,0 +1,102 @@
+For more information look at pull request number [#113](https://github.com/vmware/precaution/pull/113)
+
+
+### Check run and check suite events
```
As above, didn't we agree to use theGitHub event names for the section titles?

MVrachev replied:
Yes, my mistake.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+Precaution needs to subscribe to a number of events
+to satisfy different use cases.
+
+The events are separated into two groups `PullRequestEvent` and `CheckRunEvent/CheckSuiteEvent`.
```
```suggestion
The events are separated into two groups `PullRequestEvent` and `CheckRunEvent and CheckSuiteEvent`.
```

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+
+### PullRequestEvent
+
+Official GitHub documentation: [PullRequest event](https://developer.github.com/v3/activity/events/types/#pullrequestevent)
```
Is it "PullRequest event" or "PullRequestEvent"?

MVrachev replied:
Yep, my mistake.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+
+A user had already opened a pull request and is at the pull request page.
+He navigates to the "Checks" tab.
+If the Precaution check had failed he will see "Re-run all checks", "Re-run" and
```
```suggestion
If the Precaution check failed he will see "Re-run all checks", "Re-run" and
```

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+#### `check_suite.rerequested`
+
+- __occurs__: when you click "Re-run all checks"
+or "Re-run failed checks" buttons*
```
```suggestion
or "Re-run failed checks" buttons
```

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+
+## Events workflows
+
+This paragraph doesn't aim to explain in details how Precaution works.
```
```suggestion
This paragraph doesn't aim to explain in detail how Precaution works.
```

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+
+- __triggers__: start of a new Precaution scan
+
+For more information look at pull request number [#113](https://github.com/vmware/precaution/pull/113)
```
```suggestion
For more information look at pull request [Fix for "across the forks" PRs problems](https://github.com/vmware/precaution/pull/113)
```

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+
+- __triggers__: start of a new Precaution scan
+
+## Events workflows
```
```suggestion
## Event workflows
```

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+- [CheckSuiteEvent](https://developer.github.com/v3/activity/events/types/#checksuiteevent)
+
+A user had already opened a pull request and is at the pull request page.
+He navigates to the "Checks" tab.
```
```suggestion
They navigate to the "Checks" tab.
```

ericwb replied:
Still shows "he".  Is the user always a "he"?

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+
+A user had already opened a pull request and is at the pull request page.
+He navigates to the "Checks" tab.
+If the Precaution check failed he will see "Re-run all checks", "Re-run" and
```
```suggestion
If the Precaution check failed they will see "Re-run all checks", "Re-run" and
```

ericwb replied:
Still shows "he".  Is the user always a "he"?

MVrachev commented:
> Same comments as before.

Sorry, I made the changes but I was distracted by something else and I didn't push them.

ericwb commented on docs/events_handling.md:
```diff
@@ -0,0 +1,101 @@
+- [CheckSuiteEvent](https://developer.github.com/v3/activity/events/types/#checksuiteevent)
+
+A user had already opened a pull request and is at the pull request page.
+They navigates to the "Checks" tab.
```
```suggestion
They navigate to the "Checks" tab.
```
